jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/profiles/ProfileCredentialProviderFactory.kt

```diff
@@ -3,26 +3,293 @@
+import software.amazon.awssdk.http.SdkHttpClient
+import software.amazon.awssdk.profiles.Profile
+import software.amazon.awssdk.profiles.ProfileProperty
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.sts.StsClient
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest
+import software.aws.toolkits.core.ToolkitClientManager
+import software.aws.toolkits.core.credentials.ToolkitCredentialsIdentifier
+import software.aws.toolkits.core.region.AwsRegion
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.core.credentials.CorrectThreadCredentialsProvider
+import software.aws.toolkits.jetbrains.core.credentials.CredentialIdentifierChange
 import software.aws.toolkits.jetbrains.core.credentials.CredentialProviderFactory
-import software.aws.toolkits.jetbrains.core.region.AwsRegionProvider
-
-class ProfileCredentialProviderFactory : CredentialProviderFactory {
-    override fun createToolkitCredentialProviderFactory(manager: ToolkitCredentialsProviderManager) = ProfileToolkitCredentialsProviderFactory(
-        AwsSdkClient.getInstance().sdkHttpClient,
-        AwsRegionProvider.getInstance(),
-        manager,
-        profileWatcher
-    )
+import software.aws.toolkits.jetbrains.core.credentials.CredentialsChangeListener
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import software.aws.toolkits.jetbrains.utils.createNotificationExpiringAction
+import software.aws.toolkits.jetbrains.utils.createShowMoreInfoDialogAction
+import software.aws.toolkits.jetbrains.utils.notifyError
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.resources.message
+import java.util.function.Supplier
+
+const val DEFAULT_PROFILE_ID = "profile:default"
+
+private const val PROFILE_FACTORY_ID = "profileCredentialProviderFactory"
+
+private class ProfileCredentialsIdentifier(internal val profileName: String) : ToolkitCredentialsIdentifier() {
+    override val id = "profile:$profileName"
+    override val displayName = message("credentials.profile.name", profileName)
+    override val factoryId = PROFILE_FACTORY_ID
+}
+
+class ProfileCredentialProviderFactory : CredentialProviderFactory, Disposable {
+
+    private val profileWatcher = ProfileWatcher(this)
+    private val profileHolder = ProfileHolder()
+
+    override val id = PROFILE_FACTORY_ID
+
+    override fun setUp(credentialLoadCallback: CredentialsChangeListener) {
+        // Load the initial data, then start the background watcher
+        loadProfiles(credentialLoadCallback, true)
 
-    companion object {
-        val profileWatcher = ProfileWatcher().also {
-            it.start()
-            // TODO: Scope this better in the cred manager refactor
-            Disposer.register(ApplicationManager.getApplication(), it)
+        profileWatcher.start {
```

Comment:
nit: looks like it will call loadProfiles again if you don't dive into `ProfileWatcher` maybe a comment that it's a callback would be good

Reply:
How about this?

```
        profileWatcher.start(callback = {
            loadProfiles(credentialLoadCallback, false)
        })
```

Reply:
looks good